### ISSUE TITLE
Add the ability for AnimationPlayerState to automatically transition to other states when the animation ends.

### ISSUE THREAD

**Issue by shanayyy** [derkork/godot-statecharts#131]
A common use case for `AnimationPlayerState` is transitioning to a different state at the end of an animation, e.g. returning to the idle state after an attack animation. Currently we have to handle this transition manually. This requires tracking the completion states of different animations and mapping them to transitions.

To make `AnimationPlayerState` easier to use, I suggest adding a `Next State` property to it. If a state is assigned to it, when the `AnimationPlayer` emits the `animation_finished` signal, if the `anim_name` is equal to the `animation_name` of the `AnimationPlayerState`, and the `AnimationPlayerState` is active, it automatically transitions to the state specified by the `Next State`.
<img width="411" alt="01" src="https://github.com/derkork/godot-statecharts/assets/13399488/f5dd2a20-abce-454e-8b13-11e76f5eed91">

**Comment by derkork** [comment]
The animation control is a rather tricky part of the library. I added it as an experimental feature a while ago but animation control can become very complex and this implementation isn't going to work for all cases. 

In many cases (I would even go so far as to say in most cases) the animation length correlates with some game mechanic and has direct influence on how the game feels. E.g. firing animation of a gun needs to be properly aligned to the gun's firing rate, the strike animation of a sword strike needs to correlate with the intended fighting speed of the player or an invulnerability or stun animation is only as long as the designer said these effects would last. 

In these cases I think it is _not good to synchronize_ the animation state back to the state chart because _the game design will dictate how long these animations play_. So the state chart would not "wait" until the animation is finished - but rather control the timing itself. This way the animations are only visualizations of the game state to the player but have no impact on gameplay. For this the _AnimationPlayerState_ works fine, because when you work like this, animations are basically fire and forget.  So in your concrete example, say you have a series of attacks that should play in sequence you can do a ...

![image](https://github.com/user-attachments/assets/da9b6fd8-52b4-4454-8bcb-0ebca79187b1)

... and just need to make sure the animation lengths match the designed timings in your state chart. 

Then no manual signal mapping is needed. And for this example I think this is the better way to do it as you can tweak the timings until everything feels right and when you're done you just change the animation one time to have the proper length rather than having to modify the animation every time you want to tweak the attack timing. 

However there are also cases when you want callbacks from your animation (either through a call track or just by listening to the signals of the animation player / animation tree). This is mainly the case for purely visual things (e.g. UI animations) but there are also some use cases for quick time events. However these are usually very specific to the game and I'm having a really hard time coming up with a solution that is universally useful rather than catering only for a specific use case, like the suggested solution with the _Next State_ flag. Also I really don't want to mix transitions with states which this implementation would do.  If you want to keep this approach, I'd suggest you do it this way:

1. write a small script which handles the animation finished event from the animation player and sends an event to the state chart that corresponds to the current `anim_name`, e.g.

```gdscript
func on_animation_player_animation_finished(anim_name:StringName):
    _state_chart.send_event("animation_finished_" + anim_name)
```

2. then in your state chart just add transitions that correspond to the events:

![image](https://github.com/user-attachments/assets/3c664a4b-9902-40b1-ad62-20ede9dc964b)

That's 2 lines of code to get the specific behaviour you want and you can just use the built-in transitions to do the rest of the work. Would that be a workable solution for you?

**Comment by shanayyy** [comment]
Thank you for taking the time to write this, it's easy to understand and i agree with you. In fact I'm using your solution.It works really well and doesn't require much boilerplate code, as I've found that I only really need to create one animation_finished event for all animations. 
<img width="421" alt="image" src="https://github.com/user-attachments/assets/66ac0414-946e-4226-8f6e-e98f21cb0d86">
I have tried all the state machine plugins in godot assetlib, and finally found that only this plugin really solved my problem, but I found that ExpressionGuard seemed too slow, so that Auto Transition sometimes missed the input check, so I now send input events every frame instead. This makes the[ condition state](https://statecharts.dev/glossary/condition-state.html) pattern less useful.

**Comment by derkork** [comment]
I'm not sure what you mean by "expression guard is to slow". A guard never _triggers_ a transition it can only _prevent_  a transition from happening. There are only two ways to trigger a transition:

1. send an event
2. change an expression property

Could you elaborate on your use case? Maybe there is a nice solution for that that  doesn't require sending events every frame.

**Comment by shanayyy** [comment]
Haha, that's a typo. I mean "too slow". I drew a diagram to show what I want to do:
<img width="1515" alt="image" src="https://github.com/user-attachments/assets/d7e2d6e3-c520-4664-99ba-26d90211d823">

**Comment by derkork** [comment]
I don't think this setup will work because the resolver doesn't actually wait to resolve. When the animations finish the resolver state is entered and the resolver state will immediately  walk through all the transitions. If you haven't pressed the key yet, it will not wait but just straight evaluate the other guards and pick the first one matching.  One example (TI = Time index in milliseconds):

- TI=0  we are in Combo1, the animation finishes
  - Resolver is entered
  - We assume speed is 0, so we are now in Idle
- TI=50 - attack is pressed, nothing happens anymore because we are already in idle.

Or if you press the key slightly before the animation ends:

- TI=0, we are in Combo1, player presses attack key
   - We enter Combo2
- TI=25, we are in Combo2, player pressess attack key
- TI=75, animation ends , we return to Resolver
  -  Now the attack key was actually pressed within the last 100ms, so this time we go back to Combo1

So this is not caused by the execution speed of the evaluations, they are always fully evaluated within a single frame.

**Comment by shanayyy** [comment]
What you describe is exactly how I would expect it to work. There are two places where I need to explain a little more.
> TI=50 - attack is pressed, nothing happens anymore because we are already in idle.

If we are in idle, pressing attack will send Input Changed event, so we will jump to Resolver immediately.

> they are always fully evaluated within a single frame.

I'm not sure about this, because I found that the delay of a single evaluation can exceed 6ms, so when there are multiple evaluations to be performed it may take more than 1 frame, but it may be that I'm doing it wrong. I will make a demo project later when I have time to see if I can reproduce this.

**Comment by valenotary** [comment]
Hello !

@derkork I was also looking for something similar, but then I read the alternate solution you suggested about the animation player just firing events... 

Do you have a recommendation on even using the state chart's animationtree/animationplayer nodes? I know you said they were experimental, but wanted to see from your experience if they actually made anything easier...

Right now I'm kind of having trouble wrapping my head around which nodes should be the sort of "source of truth" of a state. I really do love the idea of this StateChart asset, and I'm thinking that most of the time I should prioritize the StateChart states _driving_ which animation is being played. 

I like leaving the original animation timings untouched, so that just leaves mucking with the timings via the AnimationTree that will reference the animations from the AnimationPlayer (via the "Use Custom Timeline" property).

But typically (correct me if I'm wrong) -- transitioning between animations in an AnimationTree is represented ALSO by another state machine (AnimationStateMachine). Naively, I'm just thinking that upon entering a state via an event, then it's my responsibility to always call the corresponding state_machine.travel("animation_state"). I have an inkling annoying to repeat across all event signals, but I'm gonna guess this just gives me the most control and to not take it for granted?

Question is then -- should the animation transitions defined in the animation tree ever be set to autoplay? or should I let the statechart purely dictate when animations play?

**Comment by derkork** [comment]
Well in my opinion there is not definitive answer for that. I can only tell you, how I build things but that doesn't mean its the "right" or "only" way to build things.  Personally, I build everything in a way that my game code is the authoritative source of state. Everything else (animations, sounds, UI, etc..) is just driven by that. This has solved quite a few headaches for me in the past because information really only flows in one direction: from my game code outwards to "visualization". 

Also animation state machines are not the be-all and end-all of how to control animations. A lot of times its easier and more robust to control animations with a blend tree and some carefully set inputs for this tree. This way I don't have to keep two state machines in sync. Also many times there isn't even a 1:1 relationship between animation tree states and statechart states. 

There are a lot of other ways to animate things in Godot's animationtree, like oneshots, blend spaces, transitions and so on - so a lot of things can often be done much easier with these approaches rather than using a state machine for animations. 

So my recommendation would be to make it easy to trigger the desired animation behaviour from the state chart using whatever means are at your disposal while at the same time avoiding any kind of feedback from the animations to your game code if possible. 

Regarding the two animation nodes in Godot State Charts - if anything, this discussion shows that these nodes shouldn't be in the state charts project at all. They prescribe a certain way of doing things that will not work for every project and at the same time they also obstruct the view to an alternative solution because people try to make it work with these nodes, only. So I think I'm going to remove them in a future version.
